File: Array/BestTimeToBuyAndSellStock.py
'''
Source : https://oj.leetcode.com/problems/best-time-to-buy-and-sell-stock/
Author : Yuan Wang
Date   : 2018-06-02

/********************************************************************************** 
* 
* Say you have an array for which the ith element is the price of a given stock on day i.
* 
* If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), 
* design an algorithm to find the maximum profit.
*	   
**********************************************************************************/
'''

def maxProfit(prices):
	"""
	:type prices: List[int]
	:rtype: int
	"""
	max_profit, min_price = 0, float('inf')
	for price in prices:
		min_price = min(min_price, price)
		profit = price - min_price
		max_profit = max(max_profit, profit)
	return max_profit

def maxProfitB(prices):
	"""
	:type prices: List[int]
	:rtype: int
	"""
	maxCur=0
	maxSoFar=0
	for i in range(1,len(prices)):
		maxCur+=prices[i]-prices[i-1]
		maxCur=max(0,maxCur)
		maxSoFar=max(maxCur,maxSoFar)
	return maxSoFar
print(maxProfitB([7,1,5,3,6,4]))

File: Array/BestTimeToBuyAndSellStock.java
/*
Algorithm

Say the given array is:

[7, 1, 5, 3, 6, 4]

If we plot the numbers of the given array on a graph, we get:


The points of interest are the peaks and valleys in the given graph.
We need to find the largest peak following the smallest valley.
We can maintain two variables - minprice and maxprofit corresponding 
to the smallest valley and maximum profit (maximum difference between 
selling price and minprice) obtained so far respectively.

Complexity Analysis

Time complexity : O(n). Only a single pass is needed.

Space complexity : O(1). Only two variables are used.
*/

import java.lang.Math;

public class BestTimeToBuyAndSellStock{
	public static void main(String[] args){
		int array[]={7,1,5,3,6,4};
		int profit=max_profit(array);
		int profitB=maxProfit(array);
		System.out.println("The max profit is: "+profit);
		System.out.println("The max profit is: "+profitB);
	}

	public static int max_profit(int prices[]){
		int maxCur=0;
		int maxSoFar=0;
		for(int i = 1;i<prices.length;i++){
			maxCur+=prices[i]-prices[i-1];
			maxCur=Math.max(0,maxCur);
			maxSoFar=Math.max(maxCur,maxSoFar);
		}

		return maxSoFar;
	}

	public static int maxProfit(int prices[]){
		int minprice = Integer.MAX_VALUE;
        int maxprofit = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minprice)
                minprice = prices[i];
            else if (prices[i] - minprice > maxprofit)
                maxprofit = prices[i] - minprice;
        }
        return maxprofit;
	}
}